bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/TapoDeviceHttpApi.java

```diff
@@ -0,0 +1,553 @@
+        try {
+            Request httpRequest = httpClient.newRequest(url).method(HttpMethod.POST.toString());
+
+            /* set header */
+            httpRequest = setHeaders(httpRequest);
+
+            /* add request body */
+            httpRequest.content(new StringContentProvider(payload, CONTENT_CHARSET), CONTENT_TYPE_JSON);
+
+            httpRequest.timeout(TAPO_HTTP_TIMEOUT_MS, TimeUnit.MILLISECONDS).send(new BufferingResponseListener() {
+                @NonNullByDefault({})
+                @Override
+                public void onComplete(Result result) {
+                    final HttpResponse response = (HttpResponse) result.getResponse();
+                    if (result.getFailure() != null) {
+                        /* handle result errors */
+                        Throwable e = result.getFailure();
+                        if (e instanceof TimeoutException) {
+                            logger.debug("({}) sendAsyncRequest timeout'{}'", uid, e.getMessage().toString());
+                            tapoError.raiseError(ERR_CONNECT_TIMEOUT, e.getMessage().toString());
+                        } else {
+                            logger.debug("({}) sendAsyncRequest failed'{}'", uid, e.getMessage().toString());
+                            tapoError.raiseError(ERR_JSON_DECODE_FAIL, e.getMessage().toString());
+                        }
+                    } else if (response.getStatus() != 200) {
+                        logger.debug("({}) sendAsyncRequest response error'{}'", uid, response.getStatus());
+                        tapoError.raiseError(ERR_HTTP_RESPONSE, getContentAsString());
+                    } else {
+                        /* request succesfull */
+                        String rBody = getContentAsString();
+                        logger.trace("({}) requestCompleted '{}'", uid, rBody);
+                        /* handle result */
+                        switch (command) {
+                            case DEVICE_CMD_SETINFO:
+                                handleSuccessResponse(rBody);
+                                break;
+                            case DEVICE_CMD_GETINFO:
+                                handleDeviceResult(rBody);
+                                break;
+                        }
+                    }
+                    /* request error */
+                    if (tapoError.hasError()) {
+                        handleErrorResult(tapoError);
+                    }
+                }
+            });
+        } catch (Exception e) {
+            tapoError.raiseError(e);
+        }
+        if (tapoError.hasError()) {
+            handleErrorResult(tapoError);
+        }
+    }
+
+    protected void handleSuccessResponse(String responseBody) {
+    }
+
+    protected void handleDeviceResult(String responseBody) {
+    }
+
+    protected void handleErrorResult(TapoErrorHandler te) {
+    }
+
+    protected void handleErrorResult(Exception ex) {
+    }
```

Comment:
What's the purpose of these?

Reply:
will normaly be delegated to extension-classes (TapoDeviceConnector)